src/rascal/math/kvec_generator.hh

```diff
@@ -0,0 +1,229 @@
+ *        lattice that lie within a ball of a specified cutoff radius.
+ *
+ * Copyright  2021  Andrea Grisafi, Kevin Kazuki Huguenin-Dumittan, COSMO(EPFL)
+ *
+ * Rascal is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3, or (at
+ * your option) any later version.
+ *
+ * Rascal is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this software; see the file LICENSE. If not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+#ifndef SRC_RASCAL_MATH_KVEC_GENERATOR_HH_
+#define SRC_RASCAL_MATH_KVEC_GENERATOR_HH_
+
+#include "rascal/math/utils.hh"
+
+#include <Eigen/Dense>
+#include <Eigen/Eigenvalues>
+
+namespace rascal {
+  namespace math {
+    /**
+     * Produce the k-vectors (= reciprocal space or Fourier vectors)
+     * needed for the k-space implementation of LODE and SOAP.
+     *
+     * More specifically, if b1, b2 and b3 are basis vectors of a reciprocal
+     * lattice, this class is used to get all vectors of the form:
+     * \f{equation}{k = n_1b_1 + n_2b_2 + n_3b_3,\f}
+     * where n1,n2,n3 are integers, which are inside a ball of a given
+     * cutoff radius \f$k_\mathrm{cut}\f$.
+     *
+     * Note that for any radius, the origin \f$k=(0,0,0)\f$ is always in
+     * the ball and is the first entry of the returned vectors.
+     * For applications where k=0 should be excluded, there is the option
+     * to do so.
+     * For any other k-vector k, symmetry typically allows us to group the
+     * two vectors k and -k together. Thus, only one out of each such pair
+     * is returned as a representant.
+     *
+     */
+    class Kvectors {
+     private:
+      /* The two quantities that completely determine the mathematical
+       * problem are the cutoff radius in reciprocal space as well as
+       * the three basis vectors of the reciprocal cell. These are provided
+       * upon initialization (either directly or indirectly) by the user:
+       */
+      double kcutoff{};
+      Eigen::Matrix3d basisvecs{};
+
+      /* Quantities that can be returned by this class:
+       * These are the quantities that will be used by LODE and the
+       * reciprocal space implementation of SOAP. They are:
+       * - the total number numvectors of vectors within the cutoff,
+       * - the numvectors times 3 Eigen matrix containing those vectors,
+       * - the Eigen vector of size numvectors containing the norms
+       *   of the found vectors.
+       */
+      size_t numvectors = 0;
+      Matrix_t kvecs{};
+      Vector_t kvecnorms{};
```

Comment:
Instead of duplicating the information on the number of vectors, you could remove the `numvectors` member, and use `kvecsnorms.size()` to implement `get_numvectors()`.